src/content/en/updates/2019/06/image-support-for-async-clipboard.md

```diff
@@ -0,0 +1,299 @@
+  console.log('Image copied.');
+} catch(e) {
+  console.error(e, e.message);
+}
+```
+
+### Paste: Reading an image from the clipboard
+
+Similar to `write()`, the `readText()` method is just a convenience method for `read()`.
+It is likewise asynchronous and Promise-based.
+
+As a first step, you need to obtain the list of `ClipboardItem`s that you then need to iterate over.
+Everything is asynchronous code, so remember to use the
+[`for … of`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...of)
+loop as it copes well with `await`ing asynchronous results.
+
+Each `ClipboardItem` can hold its contents in different types, so next,
+you need to iterate over the list of types, again using a `for … of` loop.
+For each type, you call the `getType()` method with the current type as an argument
+to obtain the corresponding image `Blob`.
+As before, this code is future-proof and not tied to images.
+
+```js
+try {
+  await navigator.permissions.request({name: 'clipboard-read'});
+  const clipboardItems = await navigator.clipboard.read();
+  for (const clipboardItem of clipboardItems) {
+    try {
+      for (const type of clipboardItem.types) {
+        const blob = await clipboardItem.getType(type);
+        console.log(URL.createObjectURL(blob));
+      }
+    } catch (e) {
+      console.error(e, e.message);
+    }
+  }
+} catch (e) {
+  console.error(e, e.message);
+}
+```
+
+### Custom paste handler
+
+If you want to dynamically handle paste events, which most likely you want to do,
+the code snippet below has all the steps you need.
+It is actually just the snippet from above, but wrapped in an event listener listening on `paste`
+that [prevents](https://developer.mozilla.org/en-US/docs/Web/API/Event/preventDefault)
+the default paste behavior.
+
+```js
+document.addEventListener('paste', async (e) => {
+  e.preventDefault();
+  try {
+    await navigator.permissions.request({name: 'clipboard-read'});
+    const clipboardItems = await navigator.clipboard.read();
+    for (const clipboardItem of clipboardItems) {
+      try {
+        for (const type of clipboardItem.types) {
+          const blob = await clipboardItem.getType(type);
+          console.log(URL.createObjectURL(blob), type);
+        }
+      } catch (e) {
+        console.error(e, e.message);
+      }
+    }
+  } catch (e) {
+    console.error(e, e.message);
+  }
+});
+```
+
+### Custom copy handler
+
+Rather than manually copying items in the clipboard, which is a little involved
+as you need to deal with `Blob`s and wrap them in `ClipboardItem`s manually,
+leveraging the `copy` event of the browser is easier.
+Again you need to prevent the default behavior, and can then just write the clipboard data
+that is conveniently already
+
+```js
+document.addEventListener('copy', async (e) => {
+  e.preventDefault();
+  try {
+    await navigator.permissions.request({name: 'clipboard-write'});
+    await navigator.clipboard.write(e.clipboardData.items);
```

Comment:
I don't believe that this line is valid. e.clipboardData.items seems to be a DataTransfer object, and not a ClipboardItem array. navigator.clipboard.write only accepts ClipboardItems, which support async operations. As DataTransfer isn't async-friendly, unfortunately navigator.clipboard.write no longer supports DataTransfer items.

Could we please expand this to be the unfortunately quite long-form version using ClipboardItem?

Reply:
@dway123 I'm not sure what I have in https://github.com/google/WebFundamentals/blob/d887f81f0388b95821f15c7443e26421b3b1396b/src/content/en/updates/2019/06/image-support-for-async-clipboard.md is correct (actually, I am sure it's wrong, but I don't see how to fix it):

```js
document.addEventListener('copy', async (e) => {
  e.preventDefault();
  try {
    await navigator.permissions.request({name: 'clipboard-write'});
    for (const item of e.clipboardData.items) {
      await navigator.clipboard.write(new ClipboardItem(Object.defineProperty({}, item.type, {
        value: item,
        enumerable: true
      })));
    }
    console.log('Image copied.');
  } catch(e) {
    console.error(e, e.message);
  }
});
```

When I try right-clicking any PNG image on a test page and select "Copy image", the handler should fire, but it doesn't. Not sure how I am holding it wrong.

`e.clipboardData.items` is a `DataTransferItemList`, which according to the [spec](https://w3c.github.io/clipboard-apis/#to-write-content-to-clipboard) is what is needed. Can you help out, please?

Reply:
Oh, and please ping @petele to get this published once it's working, as I will be OoO until July 23. Thanks!

Reply:
In the example above, an array of ClipboardItems is expected, so `await navigator.clipboard.write([new ClipboardItem(..)]);`

@petele

Reply:
Perfect, yep - I updated that one already!